[MM 22957] webapp a11y: fix sso btns focus issue

saideepesh000 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->

This PR fixes the ally issue of sso buttons on login page

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
fixes #22957 

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

https://user-images.githubusercontent.com/43727167/236675329-475702c3-80e0-4158-823b-ab1b386ff6a1.mov



#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
[MM 22957] webapp a11y: fix sso btns focus issue
```

mattermost-build commented:
Hello @saideepesh000,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

mattermost-build commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

pvev commented on webapp/channels/src/components/external_login_button/external_login_button.tsx:
```diff
@@ -1,18 +1,21 @@
 
 import './external_login_button.scss';
 
+
```
unnecessary blank line, please remove

pvev commented on webapp/channels/src/components/external_login_button/external_login_button.tsx:
```diff
@@ -22,12 +25,17 @@ const ExternalLoginButton = ({
         style={style}
+        onFocus={onFocus}
+        onBlur={onBlur}
+
```
unnecessary blank line, please remove

pvev replied:
not sure why is not getting fixed, it not a blocker but is a good practice to not add unnecessary lines, specially before closing tags.

pvev commented on webapp/channels/src/components/login/login.tsx:
```diff
@@ -121,6 +121,8 @@ const Login = ({onCustomizeHeader}: LoginProps) => {
     const [hasError, setHasError] = useState(false);
     const [isMobileView, setIsMobileView] = useState(false);
+    const [isExternalBtnFocused, setIsExternalBtnFocused] = useState<boolean>(false);
+
```
unnecessary blank line, please remove

pvev commented on webapp/channels/src/components/login/login.tsx:
```diff
@@ -129,22 +131,31 @@ const Login = ({onCustomizeHeader}: LoginProps) => {
     const enableSignInWithEmail = EnableSignInWithEmail === 'true';
     const enableSignInWithUsername = EnableSignInWithUsername === 'true';
     const enableSignUpWithEmail = EnableSignUpWithEmail === 'true';
-    const enableSignUpWithGitLab = EnableSignUpWithGitLab === 'true';
```
This logic needs to remain the way it was before, we can not assign this logic to `true` directly. We need to check if that config value is actually enabled.

saideepesh000 replied:
damn, sorry my bad, should've check this before raising the PR

pvev commented on webapp/channels/src/components/login/login.tsx:
```diff
@@ -129,22 +131,31 @@ const Login = ({onCustomizeHeader}: LoginProps) => {
 
     const enableBaseLogin = enableSignInWithEmail || enableSignInWithUsername || ldapEnabled;
-    const enableExternalSignup = enableSignUpWithGitLab || enableSignUpWithOffice365 || enableSignUpWithGoogle || enableSignUpWithOpenId || enableSignUpWithSaml;
+    const enableExternalSignup = true;
```
same here, we can not hardcode this directly to `true`

saideepesh000 replied:
sorry, my bad, did this for testing, forgot to revert, should've done before raising PR

pvev commented on webapp/channels/src/components/login/login.tsx:
```diff
@@ -674,7 +685,9 @@ const Login = ({onCustomizeHeader}: LoginProps) => {
     const onEnterKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
         if (e.key === Constants.KeyCodes.ENTER[0]) {
-            preSubmit(e);
+            if(!isExternalBtnFocused){
```
shouldn't this be a positive evaluation? I mean, the check should be `if (isExternalBtnFocused)`

pvev commented on webapp/channels/src/components/login/login.tsx:
```diff
@@ -827,11 +840,13 @@ const Login = ({onCustomizeHeader}: LoginProps) => {
                                 </div>
                             )}
-                            {enableExternalSignup && (
+                            {true && (
```
this condition must remain as it was before. It doesn't make any sense to hardcode a bool value in a condition.

pvev commented on webapp/channels/src/components/login/login.tsx:
```diff
@@ -129,22 +131,31 @@ const Login = ({onCustomizeHeader}: LoginProps) => {
     const query = new URLSearchParams(search);
     const redirectTo = query.get('redirect_to');
 
+    const handleExternalBtnFocus = () => {
```
you can simplify this to be only one function and then use it this way:

```
const handleExternalBtnFocusBlur = (isFocused: boolean) => {
    setIsExternalBtnFocused(isFocused)
}

```

And then use it like this:

```
<ExternalLoginButton
    ...
    onFocus={() => handleExternalBtnFocusBlur(true)}
    onBlur={() => handleExternalBtnFocusBlur(false)}
    ...
/>
```

also, the function should be memoized using the useCallback function to avoid unnecessary re-renders of the child component.

pvev reviewed (changes requested):
Thank you for your contribution @saideepesh000 , the solution you propose seems like a right approach to me. I have requested some changes before approving. Please address those and let me know if you have any questions or need any help or guidance. Thanks!

saideepesh000 commented:
> Thank you for your contribution @saideepesh000 , the solution you propose seems like a right approach to me. I have requested some changes before approving. Please address those and let me know if you have any questions or need any help or guidance. Thanks!

thanks @pvev, addressed the comments

pvev reviewed (changes requested):
thanks for the quick fix to my previous change requests 🚀  @saideepesh000 . Please review my latests comments. Thanks!

saideepesh000 commented:
> thanks for the quick fix to my previous change requests 🚀 @saideepesh000 . Please review my latests comments. Thanks!

@pvev done 🚀

hmhealey commented:
Sorry for barging in a bit, but if I could actually suggest a different fix here, we don't actually need a key handler or to track focus here if we just wrap the login fields in an HTML form. A form's `submit` event will trigger whenever the user presses enter while their focus is in the form or whenever they click on a submit button (like `SaveButton` uses internally). It just looks like:
```jsx
const preSubmit = (e) => {
    e.preventDefault(); // This is important since the default browser behaviour is to make a request and reload the page

    // And so on
};

return (
    <form onSubmit={preSubmit}>
        <input id='username'/>
        <button type='submit'>{'Save'}</button>
    </form>
);
```
It also won't interfere with any other non-submit buttons in the form as well. Currently, the "Forgot your password" link and the "show password" toggle also have the same issue as the SSO buttons, but they should work just fine even if they're inside a form

saideepesh000 commented:
/update-branch

saideepesh000 commented:
/update-branch

saideepesh000 commented:
@hmhealey made the changes, please check it.

BenCookie95 reviewed (approved):
LGTM! Tested locally

hmhealey reviewed (approved):
Thanks for redoing that, @saideepesh000. I'm glad it worked how I remembered since I've only actually used `form`s a couple times. I wonder if there's other places in the app that we could use that pattern to fix similar issues too...

hmhealey commented:
/e2e-tests

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/342345) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/342345)
